cores/esp8266/core_esp8266_i2s.c

```diff
@@ -194,46 +219,94 @@ bool ICACHE_FLASH_ATTR i2s_write_lr(int16_t left, int16_t right){
   return i2s_write_sample(sample);
 }
 
+bool ICACHE_FLASH_ATTR i2s_write_lr_nb(int16_t left, int16_t right){
+  int sample = right & 0xFFFF;
+  sample = sample << 16;
+  sample |= left & 0xFFFF;
+  return i2s_write_sample_nb(sample);
+}
+
 //  END DMA
 // =========
 // START I2S
 
 
-static uint32_t _i2s_sample_rate;
+static uint32_t _i2s_sample_rate = 0;
 
 void ICACHE_FLASH_ATTR i2s_set_rate(uint32_t rate){ //Rate in HZ
   if(rate == _i2s_sample_rate) return;
   _i2s_sample_rate = rate;
-  uint32_t i2s_clock_div = (I2SBASEFREQ/(_i2s_sample_rate*32)) & I2SCDM;
-  uint8_t i2s_bck_div = (I2SBASEFREQ/(_i2s_sample_rate*i2s_clock_div*2)) & I2SBDM;
-  //os_printf("Rate %u Div %u Bck %u Frq %u\n", _i2s_sample_rate, i2s_clock_div, i2s_bck_div, I2SBASEFREQ/(i2s_clock_div*i2s_bck_div*2));
 
-  //!trans master, !bits mod, rece slave mod, rece msb shift, right first, msb right
-  I2SC &= ~(I2STSM | (I2SBMM << I2SBM) | (I2SBDM << I2SBD) | (I2SCDM << I2SCD));
-  I2SC |= I2SRF | I2SMR | I2SRSM | I2SRMS | ((i2s_bck_div-1) << I2SBD) | ((i2s_clock_div-1) << I2SCD);
+  // Find closest divider
+  uint32_t cpu_freq = ets_get_cpu_frequency() * 1000000L;
+  int bestfreq = 0;
+  uint32_t i2s_clkm_div, i2s_bck_div;
+
+  // CLK_I2S = CPU_FREQ / I2S_CLKM_DIV_NUM
+  // BCLK = CLK_I2S / I2S_BCK_DIV_NUM
+  // WS = BCLK/ 2 / (16 + I2S_BITS_MOD)
+  // Note that I2S_CLKM_DIV_NUM must be >5 for I2S data
+  // I2S_CLKM_DIV_NUM - 5 - 63
+  // I2S_BCK_DIV_NUM - 2 - 63
+  for (int bckdiv = 2; bckdiv < 64; bckdiv++) {
+    for (int clkmdiv = 5; clkmdiv < 64; clkmdiv++) {
+      uint32_t testfreq = cpu_freq / (bckdiv * clkmdiv * 32);
```

Comment:
I ran a test program
```
#include <iostream>
#include <cstdint>

#define BASEFREQ (160L * 1000L * 1000L)

void i2s_set_rate(int rate, bool bits_mod) {
	//Find closest divider 
	int bestclkmdiv, bestbckdiv, bestbits, bestfreq=0;
	int tstfreq;
	/*
	   CLK_I2S = 160MHz / I2S_CLKM_DIV_NUM
	   BCLK = CLK_I2S / I2S_BCK_DIV_NUM
	   WS = BCLK/ 2 / (16 + I2S_BITS_MOD)
	   Note that I2S_CLKM_DIV_NUM must be >5 for I2S data
	   I2S_CLKM_DIV_NUM - 5-63
	   I2S_BCK_DIV_NUM - 2-63

	   We also have the option to send out more than 2x16 bit per sample. Most I2S codecs will
	   ignore the extra bits and in the case of the 'fake' PWM/delta-sigma outputs, they will just lower the output
	   voltage a bit, so we add them when it makes sense. Some of them, however, won't accept it, that's
	   why we have the option not to do this.
	 */
	for (int bckdiv=2; bckdiv < 64; bckdiv++) {
		for (int clkmdiv = 5; clkmdiv < 64; clkmdiv++) {
			for (int bits = 16; bits < (bits_mod ? 17 : 20); bits++) {
				tstfreq=BASEFREQ/(bckdiv*clkmdiv*bits*2);
				if (std::abs(rate-tstfreq)<std::abs(rate-bestfreq)) {
					bestfreq=tstfreq;
					bestclkmdiv=clkmdiv;
					bestbckdiv=bckdiv;
					bestbits=bits;
				}
			}
		}
	}

	std::cout
		<< "ReqRate " << rate << ", "
		<< "MclkDiv " << bestclkmdiv << ", "
		<< "BckDiv " << bestbckdiv << ", "
		<< "Bits " << bestbits << ", "
		<< "Freq " << (BASEFREQ / (bestbckdiv * bestclkmdiv * bestbits * 2)) << ", "
		<< "diff " << (rate - (BASEFREQ / (bestbckdiv * bestclkmdiv * bestbits * 2)))
		<< std::endl;
}

#define I2SCDM (0x3f)
#define I2SBDM (0x3f)
#define I2SBASEFREQ (12L * 1000L * 1000L)

void i2s_set_rate_static(int rate) {
	uint32_t i2s_clock_div = (I2SBASEFREQ/(rate*32)) & I2SCDM;
	uint32_t i2s_bck_div = (I2SBASEFREQ/(rate*i2s_clock_div*2)) & I2SBDM;

	std::cout
		<< "ReqRate " << rate << ", "
		<< "MclkDiv " << i2s_clock_div << ", "
		<< "BckDiv " << i2s_bck_div << ", "
		<< "Freq " << (I2SBASEFREQ / (i2s_clock_div * i2s_bck_div * 2)) << ", "
		<< "diff " << (rate - (I2SBASEFREQ / (i2s_clock_div * i2s_bck_div * 2)))
		<< std::endl;
}

int main() {
	int rates[] = { 8000, 11025, 16000, 22050, 32000, 44100, 48000, 88200, 96000 };
	for (int i = 0;i < sizeof(rates) / sizeof(*rates);++i) {
		std::cout << "mp3 demo (bits mod)   : ";
		i2s_set_rate(rates[i], true);
		std::cout << "mp3 demo (no bits mod): ";
		i2s_set_rate(rates[i], false);
		std::cout << "esp8266/Arduino       : ";
		i2s_set_rate_static(rates[i]);
		std::cout << std::endl;
	}
}
```
and got these result:
```
mp3 demo (bits mod)   : ReqRate 8000, MclkDiv 25, BckDiv 25, Bits 16, Freq 8000, diff 0
mp3 demo (no bits mod): ReqRate 8000, MclkDiv 25, BckDiv 25, Bits 16, Freq 8000, diff 0
esp8266/Arduino       : ReqRate 8000, MclkDiv 46, BckDiv 16, Freq 8152, diff -152

mp3 demo (bits mod)   : ReqRate 11025, MclkDiv 35, BckDiv 13, Bits 16, Freq 10989, diff 36
mp3 demo (no bits mod): ReqRate 11025, MclkDiv 31, BckDiv 13, Bits 18, Freq 11028, diff -3
esp8266/Arduino       : ReqRate 11025, MclkDiv 34, BckDiv 16, Freq 11029, diff -4

mp3 demo (bits mod)   : ReqRate 16000, MclkDiv 52, BckDiv 6, Bits 16, Freq 16025, diff -25
mp3 demo (no bits mod): ReqRate 16000, MclkDiv 49, BckDiv 6, Bits 17, Freq 16006, diff -6
esp8266/Arduino       : ReqRate 16000, MclkDiv 23, BckDiv 16, Freq 16304, diff -304

mp3 demo (bits mod)   : ReqRate 22050, MclkDiv 57, BckDiv 4, Bits 16, Freq 21929, diff 121
mp3 demo (no bits mod): ReqRate 22050, MclkDiv 38, BckDiv 5, Bits 19, Freq 22160, diff -110
esp8266/Arduino       : ReqRate 22050, MclkDiv 17, BckDiv 16, Freq 22058, diff -8

mp3 demo (bits mod)   : ReqRate 32000, MclkDiv 52, BckDiv 3, Bits 16, Freq 32051, diff -51
mp3 demo (no bits mod): ReqRate 32000, MclkDiv 49, BckDiv 3, Bits 17, Freq 32012, diff -12
esp8266/Arduino       : ReqRate 32000, MclkDiv 11, BckDiv 17, Freq 32085, diff -85

mp3 demo (bits mod)   : ReqRate 44100, MclkDiv 57, BckDiv 2, Bits 16, Freq 43859, diff 241
mp3 demo (no bits mod): ReqRate 44100, MclkDiv 19, BckDiv 5, Bits 19, Freq 44321, diff -221
esp8266/Arduino       : ReqRate 44100, MclkDiv 8, BckDiv 17, Freq 44117, diff -17

mp3 demo (bits mod)   : ReqRate 48000, MclkDiv 52, BckDiv 2, Bits 16, Freq 48076, diff -76
mp3 demo (no bits mod): ReqRate 48000, MclkDiv 49, BckDiv 2, Bits 17, Freq 48019, diff -19
esp8266/Arduino       : ReqRate 48000, MclkDiv 7, BckDiv 17, Freq 50420, diff -2420

mp3 demo (bits mod)   : ReqRate 88200, MclkDiv 19, BckDiv 3, Bits 16, Freq 87719, diff 481
mp3 demo (no bits mod): ReqRate 88200, MclkDiv 24, BckDiv 2, Bits 19, Freq 87719, diff 481
esp8266/Arduino       : ReqRate 88200, MclkDiv 4, BckDiv 17, Freq 88235, diff -35

mp3 demo (bits mod)   : ReqRate 96000, MclkDiv 26, BckDiv 2, Bits 16, Freq 96153, diff -153
mp3 demo (no bits mod): ReqRate 96000, MclkDiv 7, BckDiv 7, Bits 17, Freq 96038, diff -38
esp8266/Arduino       : ReqRate 96000, MclkDiv 3, BckDiv 20, Freq 100000, diff -4000
```

looks like the codes from mp3 demo does produce better `bck_div` and `clkm_div`.